### PR TITLE
Ticket1592 scientific notation

### PR DESCRIPTION
**Pull request by IsabellaRey**
Numbers <0.001 or >=1000000 are represented in scientific notation (doubles only), taking into account also the PREC field of the PV.

I have also added new entries in the manual system tests spreadsheet (tests 21-7 to 21-23)

Note I have created some new PVs in the SIMPLE IOC, to make testing easier (see ISISComputingGroup/EPICS-SimpleIoc#1 )

This solves for example issues with small pressures being displayed as 0 in the GUI.

### To test

ISISComputingGroup/IBEX#1592

### Acceptance criteria

Create blocks using the PVs in SIMPLE IOC. For this purpose there are two ai records (with PREC 3 and 5, though you can change it at runtime) and a LONG. There is also a string record that could be used.

Test changing the PV values, some examples are in the table in the ticket itself.
Please also check edge cases: values just above/below/equal to thresholds (0.001 and 1000000), negative values and 0, values needing rounding.

You can also switch to real instruments and look at how their blocks are formatted.

---

#### Code Review

- [x] Is the code of an acceptable quality?
- [x] Does the code conform to the [coding standards](https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/GUI-Coding-Conventions)? Is it well structured with small focussed classes/methods/functions?
- [x] Are there unit tests in place? Are the unit tests small and test the a class in isolation?
- [ ] Are there [system tests](https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/System-Testing-with-RCPTT) in place? Do they test a minimal set of functionality and leave the gui as close as possible to its original state?
- [x] Have the changes been documented in the [release notes](https://github.com/ISISComputingGroup/IBEX/wiki/ReleaseNotes_Dev). If so, do they describe the changes appropriately?

### Functional Tests

- [x] Do changes function as described? Add comments below that describe the tests performed.
- [x] How do the changes handle unexpected situations, e.g. bad input?
- [x] Has developer documentation been updated if required?

### Final Steps
- [x] Reviewer has moved the [release notes](https://github.com/ISISComputingGroup/IBEX/wiki/ReleaseNotes_Dev) entry for this ticket in the "Changes merged into master" section

### CODE REVIEW

File: base/uk.ac.stfc.isis.ibex.epics/src/uk/ac/stfc/isis/ibex/epics/conversion/ExponentialOnThresholdFormat.java [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+ * the default format.
+ */
+public class ExponentialOnThresholdFormat extends NumberFormat {
+
+    private static final String EXPONENTIAL_PATTERN = "0.####E0";
+    private static final double SMALL_NUMBER_THRESHOLD = 0.001;
+    private static final double BIG_NUMBER_THRESHOLD = 1000000;
+    private static final int INTEGER_EXPO_FRACTIONAL_DIGITS = 4;
+
+    private NumberFormat defaultFormat;
```
Comment (AdrianPotter) [inline base/uk.ac.stfc.isis.ibex.epics/src/uk/ac/stfc/isis/ibex/epics/conversion/ExponentialOnThresholdFormat.java]: This is only set by the constructor. Can it be made final?

File: base/uk.ac.stfc.isis.ibex.epics/src/uk/ac/stfc/isis/ibex/epics/conversion/ExponentialOnThresholdFormat.java [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+    public Number parse(String source, ParsePosition parsePosition) {
+        return defaultFormat.parse(source, parsePosition);
+    }
+
+    private boolean needsExponentialFormat(double number) {
+        if (number == 0.0) {
+            return false;
+        }
+
+        return Math.abs(number) < SMALL_NUMBER_THRESHOLD || Math.abs(number) >= BIG_NUMBER_THRESHOLD;
```
Comment (AdrianPotter) [inline base/uk.ac.stfc.isis.ibex.epics/src/uk/ac/stfc/isis/ibex/epics/conversion/ExponentialOnThresholdFormat.java]: I find this reads a little unusually. You combine a standalone if clause with a compound boolean. Why not do everything as an if-else, or everything as a single boolean? The latter is bordering on overly complex:

```
return (Math.abs(number) < SMALL_NUMBER_THRESHOLD || Math.abs(number) >= BIG_NUMBER_THRESHOLD) && number != 0.0;
```

I slightly favour:

```
if (number == 0.0) {
    return false;
} else if (Math.abs(number) < SMALL_NUMBER_THRESHOLD || Math.abs(number) >= BIG_NUMBER_THRESHOLD) {
    return true;
} else {
    return false;
}
```

File: base/uk.ac.stfc.isis.ibex.epics/src/uk/ac/stfc/isis/ibex/epics/conversion/VTypeDefaultFormatter.java [hunk 3]
```diff
@@ -59,8 +73,10 @@ public String convert(VType value) throws ConversionException {
 		}
 	}
 	
-	public String asString(VNumber vnum) {
-		return vnum.getFormat().format(vnum.getValue());
+    private String asString(VNumber vnum) {
+        NumberFormat vnumFormat = vnum.getFormat();
+        NumberFormat exponentialWhenNeededFormat = new ExponentialOnThresholdFormat(vnumFormat);
+        return exponentialWhenNeededFormat.format(vnum.getValue());
```
Comment (AdrianPotter) [inline base/uk.ac.stfc.isis.ibex.epics/src/uk/ac/stfc/isis/ibex/epics/conversion/VTypeDefaultFormatter.java]: Could be done as:

```
return (new ExponentialOnThresholdFormat(vnum.getFormat())).format(vnum.getValue());
```

Personally I don't think this reduces readability given the already long variable names. At the very least I think `vnumFormat` is redundant and can be replaced directly with `vnum.getFormat()`

File: base/uk.ac.stfc.isis.ibex.epics/src/uk/ac/stfc/isis/ibex/epics/conversion/VTypeDefaultFormatter.java [hunk 4]
```diff
@@ -1,36 +1,50 @@
...
 
     private static final int ARRAY_ELEMENTS_TO_DISPLAY = 10;
 
+    /**
+     * Returns a converter for formatting the input and its units.
+     */
     public final Converter<T, String> withUnits = new WithUnits<T>();
+
+    /**
+     * Returns a converted for formatting the input without units.
```
Comment (AdrianPotter) [inline base/uk.ac.stfc.isis.ibex.epics/src/uk/ac/stfc/isis/ibex/epics/conversion/VTypeDefaultFormatter.java]: Typo: `converted` instead of `converter`

File: base/uk.ac.stfc.isis.ibex.epics.tests/src/uk/ac/stfc/isis/ibex/epics/tests/conversion/ExponentialOnThresholdFormatTest.java [hunk 5]
```diff
@@ -0,0 +1,375 @@
...
+
+        // Act-Assert
+        assertEquals(expected, format.format(value));
+    }
+
+    @Test
+    public void
+            GIVEN_value_0_1_above_lower_threshold_WHEN_formatting_with_precision_3_THEN_no_exponential_notation_used() {
+        // Arrange
+        double value = 0.1;
```
Comment (AdrianPotter) [inline base/uk.ac.stfc.isis.ibex.epics.tests/src/uk/ac/stfc/isis/ibex/epics/tests/conversion/ExponentialOnThresholdFormatTest.java]: Surely this value is absolute, not relative to the lower threshold as the name suggests. Please change the test or the test name appropriately. This applies to several subsequent tests

File: base/uk.ac.stfc.isis.ibex.epics/src/uk/ac/stfc/isis/ibex/epics/conversion/ExponentialOnThresholdFormat.java (hunk 1 above)
Comment (IsabellaRey) [inline base/uk.ac.stfc.isis.ibex.epics/src/uk/ac/stfc/isis/ibex/epics/conversion/ExponentialOnThresholdFormat.java]: Done

File: base/uk.ac.stfc.isis.ibex.epics/src/uk/ac/stfc/isis/ibex/epics/conversion/ExponentialOnThresholdFormat.java (hunk 2 above)
Comment (IsabellaRey) [inline base/uk.ac.stfc.isis.ibex.epics/src/uk/ac/stfc/isis/ibex/epics/conversion/ExponentialOnThresholdFormat.java]: Done

File: base/uk.ac.stfc.isis.ibex.epics/src/uk/ac/stfc/isis/ibex/epics/conversion/VTypeDefaultFormatter.java (hunk 4 above)
Comment (IsabellaRey) [inline base/uk.ac.stfc.isis.ibex.epics/src/uk/ac/stfc/isis/ibex/epics/conversion/VTypeDefaultFormatter.java]: Done

File: base/uk.ac.stfc.isis.ibex.epics/src/uk/ac/stfc/isis/ibex/epics/conversion/VTypeDefaultFormatter.java (hunk 3 above)
Comment (IsabellaRey) [inline base/uk.ac.stfc.isis.ibex.epics/src/uk/ac/stfc/isis/ibex/epics/conversion/VTypeDefaultFormatter.java]: Done (I went for the middle way)

File: base/uk.ac.stfc.isis.ibex.epics.tests/src/uk/ac/stfc/isis/ibex/epics/tests/conversion/ExponentialOnThresholdFormatTest.java (hunk 5 above)
Comment (IsabellaRey) [inline base/uk.ac.stfc.isis.ibex.epics.tests/src/uk/ac/stfc/isis/ibex/epics/tests/conversion/ExponentialOnThresholdFormatTest.java]: Good catch, I only added the numbers in the test names to differentiate them, hadn't realised they could be read as relative shifts!
